packages/block-library/src/table-of-contents/edit.js

```diff
@@ -15,127 +15,183 @@ import {
-	const pageIndex = useSelect(
-		( select ) => {
-			if ( ! onlyIncludeCurrentPage ) {
-				return null;
-			}
+	const {
+		__unstableMarkNextChangeAsNotPersistent,
+		replaceBlocks,
+	} = useDispatch( blockEditorStore );
 
+	/**
+	 * The latest heading data, or null if the new data deeply equals the saved headings attribute.
+	 *
+	 * Since useSelect forces a re-render when its return value is shallowly inequal to its prior call, we would be re-rendering this block every time the stores change, even if the latest headings were deeply equal to the ones saved in the block attributes.
+	 *
+	 * By returning null when they're equal, we reduce that to 2 renders: one when there are new latest headings (and so it returns them), and one when they haven't changed (so it returns null). As long as the latest heading data remains the same, further calls of the useSelect callback will continue to return null, thus preventing any forced re-renders.
+	 */
+	const latestHeadings = useSelect(
+		( select ) => {
 			const {
 				getBlockAttributes,
 				getBlockIndex,
 				getBlockName,
 				getBlockOrder,
+				getClientIdsOfDescendants,
+				__experimentalGetGlobalBlocksByName: getGlobalBlocksByName,
 			} = select( blockEditorStore );
 
-			const blockIndex = getBlockIndex( clientId );
-			const blockOrder = getBlockOrder();
-
-			// Calculate which page the block will appear in on the front-end by
-			// counting how many <!--nextpage--> tags precede it.
-			// Unfortunately, this implementation only accounts for Page Break and
-			// Classic blocks, so if there are any <!--nextpage--> tags in any
-			// other block, they won't be counted. This will result in the table
-			// of contents showing headings from the wrong page if
-			// onlyIncludeCurrentPage === true. Thankfully, this issue only
-			// affects the editor implementation.
-			let page = 1;
-			for ( let i = 0; i < blockIndex; i++ ) {
-				const blockName = getBlockName( blockOrder[ i ] );
+			// Disable reason: blocks can be loaded into a *non-post* block editor, so to avoid declaring @wordpress/editor as a dependency, we must access its store by string. When the store is not available, editorSelectors will be null, and the block's saved markup will lack permalinks.
+			// eslint-disable-next-line @wordpress/data-no-store-string-literals
+			const editorSelectors = select( 'core/editor' );
+
+			const pageBreakClientIds = getGlobalBlocksByName( 'core/nextpage' );
+
+			const isPaginated = pageBreakClientIds.length !== 0;
+
+			const tocIndex = getBlockIndex( clientId );
+
+			// Calculate the page (of a paginated post) this Table of Contents block is part of. Note that pageBreakClientIds may not be in the order they appear on the page, so we have to iterate over all of them.
+			let tocPage = 1;
+			for ( const pageBreakClientId of pageBreakClientIds ) {
+				if ( tocIndex > getBlockIndex( pageBreakClientId ) ) {
+					tocPage++;
+				}
+			}
```

Comment:
`getBlockIndex` returns the position of a block relative to its siblings. So the computation of `tocPage` breaks as soon as a ToC or a Page Break block is nested in any other block. :)

* We could place restrictions at the editor level on where/when these blocks can be inserted or moved to prevent this from ever happening (e.g. the More block has `supports: { multiple: false }` to prevent it from being inserted more than once in a post), but that seems fragile and annoying.
* We could issue user warnings or disable some functionality if we detect that any ToC or Page Break block lives anywhere else.
* Or we could implement the correct algorithm for this, which is not difficult but is a little bit more expensive. We could get the relative position of a block from the root of the tree expressed as a list of indices:

```
given the following tree:

Vertebrate
  Mammal
    Cat
    Dog
  Reptile
    Turtles
Invertebrate
  Jellyfish

path to Cat: [ 0, 0, 0 ]
path to Jellyfish: [ 1, 0 ]
```

Now, comparing positions is easy:

```
do:
  a := shift pathA
  b := shift pathB
while a == b

return a - b
```

Reply:
Wow, massive thanks for not only spotting this, but providing an elegant solution! I've implemented it now.

Notably, I tweaked the position-comparison algorithm to handle comparing the position of blocks that support nesting, even though neither the Table of Contents nor Page Break support that. I just wanted to make sure not to prepare in case someday the Page Break block gets phased out in favor of a Page block or the code gets reused elsewhere.